internal/shim-sgx/src/handler/mod.rs

```diff
@@ -174,4 +178,68 @@ impl<'a> Handler<'a> {
 
         self.ssa.gpr.rip += 2;
     }
+
+    /// Print a stack trace using the SSA registers.
+    fn print_ssa_stack_trace(&mut self) {
+        if DEBUG {
+            unsafe { self.print_stack_trace(self.ssa.gpr.rip, self.ssa.gpr.rbp) }
+        }
+    }
+
+    /// Print out `rip` relative to the shim (S) or the payload (P) base address.
+    ///
+    /// This can be used with `addr2line` and the executable with debug info
+    /// to get the function name and line number.
+    unsafe fn print_rip(&mut self, rip: u64) {
+        let exec_range = &ENARX_EXEC_START as *const u8 as u64..&ENARX_EXEC_END as *const u8 as u64;
+
+        if exec_range.contains(&rip) {
+            let rip_pie = rip - &ENARX_EXEC_START as *const u8 as u64;
+            debugln!(self, "P {:>#016x}", rip_pie);
+        } else {
+            let rip_pie = (ENCL_SIZE as u64 - 1) & rip;
+            debugln!(self, "S {:>#016x}", rip_pie);
+        }
+    }
+
+    /// Print a stack trace with the old `rbp` stack frame pointers
+    unsafe fn print_stack_trace(&mut self, rip: u64, mut rbp: u64) {
+        debugln!(self, "TRACE:");
+
+        self.print_rip(rip);
+
+        //Maximum 64 frames
+        for _frame in 0..64 {
+            if rbp == 0 || rbp & 7 != 0 {
```

Comment:
Couldn't we make this sequence of `if`s and breaks a bit nicer with a `match`?

Reply:
Something like this?

```rust
        //Maximum 64 frames
        for _frame in 0..64 {
            if rbp == 0 || rbp & 7 != 0 {
                break;
            }

            if !shim_range.contains(&rbp) {
                debugln!(self, "invalid rbp: {:>#016x}", rbp);
                break;
            }

            match rbp.checked_add(size_of::<usize>() as _) {
                None => break,
                Some(rip_rbp) => {
                    let rip = *(rip_rbp as *const u64);
                    match rip.checked_sub(1) {
                        None => break,
                        Some(0) => break,
                        Some(rip) => {
                            self.print_rip(rip);
                            rbp = *(rbp as *const u64);
                        }
                    }
                }
            }
        }
```

Reply:
Not sure that's better. I meant doing a match on `rbp` value to at least partially cover this.
Also I was really just trying to point out that the abundance of `break` statements makes this hard to read/understand and could potentially lead to bugs.
E.g. a function which performs the iteration and returns could be an improvement in my eyes, but this is OK as well.

Reply:
pushed a fixup